[quant] Statically quantized LSTM

z-a-f opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46229 [quant] Statically quantized LSTM**
* #47802 [quant][refactor] Adding fx to the 'quantization/__init__.py' to enable module import
* #47801 [quant][refactor] Moving the 'quantization/fx/utils.py' to 'quantization/utils.py'
* #47730 [quant] Add another keyword to the 'prepare_custom_config_dict' to be the same as in FX

Differential Revision: [D24270011](https://our.internmc.facebook.com/intern/diff/D24270011)

Commit:
[quant] Statically quantized LSTM

[ghstack-poisoned]

Commit:
Update on "[quant] Statically quantized LSTM"




[ghstack-poisoned]

Commit:
Update on "[quant] Statically quantized LSTM"




[ghstack-poisoned]

Commit:
Update on "[quant] Statically quantized LSTM"


Differential Revision: [D24270011](https://our.internmc.facebook.com/intern/diff/D24270011)

[ghstack-poisoned]

Commit:
Update on "[quant] Statically quantized LSTM"


Differential Revision: [D24270011](https://our.internmc.facebook.com/intern/diff/D24270011)

[ghstack-poisoned]

Commit:
Update on "[quant] Statically quantized LSTM"


Differential Revision: [D24270011](https://our.internmc.facebook.com/intern/diff/D24270011)

[ghstack-poisoned]

Commit:
Update on "[quant] Statically quantized LSTM"


Differential Revision: [D24270011](https://our.internmc.facebook.com/intern/diff/D24270011)

[ghstack-poisoned]

Commit:
Update on "[quant] Statically quantized LSTM"


Differential Revision: [D24270011](https://our.internmc.facebook.com/intern/diff/D24270011)

[ghstack-poisoned]

Commit:
Update on "[quant] Statically quantized LSTM"


Differential Revision: [D24270011](https://our.internmc.facebook.com/intern/diff/D24270011)

[ghstack-poisoned]

Commit:
Update on "[quant] Statically quantized LSTM"


Differential Revision: [D24270011](https://our.internmc.facebook.com/intern/diff/D24270011)

[ghstack-poisoned]

raghuramank100 commented on torch/nn/quantized/modules/rnn.py:
```diff
@@ -0,0 +1,288 @@
+        return x
+    return x[0]
+
+class LSTMCell(torch.nn.Module):
```
Since these are fp32 modules, lets have them in a different name-space. torch/nn/quantizable? (cc@vkuzo)

raghuramank100 commented on torch/nn/quantized/modules/rnn.py:
```diff
@@ -0,0 +1,288 @@
+        return layer
+
+
+class LSTMLayer(nn.Module):
```
Lets prefix this with an _ as this is not a user facing module.

z-a-f replied:
This should be usable by the user, but because there is no `torch.nn` equivalence of it, it is not exposed in the `__init__`. Should we still hide it?

raghuramank100 commented on torch/nn/quantized/modules/rnn.py:
```diff
@@ -0,0 +1,288 @@
+        return observed
+
+
+class _LSTMLayer(nn.Module):
```
Can we remove this class and merge it with LSTMLayer, it seems too lightweight

raghuramank100 commented on torch/nn/quantized/modules/rnn.py:
```diff
@@ -0,0 +1,288 @@
+        return observed
+
+
+class _LSTMLayer(nn.Module):
```
Might be clearer to call this _LSTMSingleLayer?

z-a-f replied:
This follows the naming convention in the `rnn.cpp`. Also, having `single` in the name implies there is a `multiple` somewhere. Should I still do it?

raghuramank100 commented on torch/nn/quantized/modules/rnn.py:
```diff
@@ -0,0 +1,288 @@
+            hidden_bw = hx_bw, cx_bw
+        hidden_fw = hx_fw, cx_fw
+        result_fw, hidden_fw = self.layer_fw(x, hidden_fw)
+        if not self.bidirectional:
```
NIT: Might be clearer to write:
```python
if self.bidirectional:
        x_reversed = x.flip(0)
        result_bw, hidden_bw = self.layer_bw(x_reversed, hidden_bw)
        result_bw = result_bw.flip(0)

        result = torch.cat([result_fw, result_bw], result_fw.dim() - 1)
        h = torch.stack([hidden_fw[0], hidden_bw[0]], 0)
        c = torch.stack([hidden_fw[1], hidden_bw[1]], 0)
        if self.batch_first:
            result.transpose_(0, 1)
       return result, (h,c)

 else:
          return result_fw, hidden_fw

raghuramank100 commented on torch/nn/quantized/modules/rnn.py:
```diff
@@ -0,0 +1,288 @@
+        hidden_fw = hx_fw, cx_fw
+        result_fw, hidden_fw = self.layer_fw(x, hidden_fw)
+        if not self.bidirectional:
+            return result_fw, hidden_fw
```
Do you need to transpose for the uni-directional case too?

raghuramank100 commented on torch/nn/quantized/modules/rnn.py:
```diff
@@ -0,0 +1,288 @@
+            self.layer_bw = _LSTMLayer(input_dim, hidden_dim, bias=bias)
+
+    def forward(self, x, hidden=None):
+        if self.batch_first:
```
Do you also need to handle packed inputs?

z-a-f replied:
I could send amendment for the packed inputs in later PRs

raghuramank100 commented on test/quantization/test_quantized_op.py:
```diff
@@ -2702,6 +2702,100 @@ def test_qrnncell(self, num_batches, input_size, hidden_size, per_channel_quant)
+        return signal, noise, snr_db
+
+    @override_qengines
+    def test_lstm(self):
```
In this case, we are testing only for (1) does the module get created correctly and (2) does it work with the quantization workflow. 
For (1), please add a test in test_quantized_module, this should check module creation, serialization, attributes etc.
For (2), please add a test in test_quantize, which should call prepare and convert and check that module swap occured as expected. 
We do not need to check numerics as we are not creating a new op here, suggest removing this test from this file.

z-a-f replied:
Numerics have to be checked, as the functionality of the LSTM is not just dependent on the linear, but also on the "wiring". If there is no unittest for numerical equivalence, it might perform a different function

raghuramank100 reviewed (changes requested):
Please see comments

Commit:
Update on "[quant] Statically quantized LSTM"


Differential Revision: [D24270011](https://our.internmc.facebook.com/intern/diff/D24270011)

[ghstack-poisoned]

Commit:
Update on "[quant] Statically quantized LSTM"


Differential Revision: [D24270011](https://our.internmc.facebook.com/intern/diff/D24270011)

[ghstack-poisoned]

Commit:
Update on "[quant] Statically quantized LSTM"


Differential Revision: [D24270011](https://our.internmc.facebook.com/intern/diff/D24270011)

[ghstack-poisoned]

Commit:
Update on "[quant] Statically quantized LSTM"


Differential Revision: [D24270011](https://our.internmc.facebook.com/intern/diff/D24270011)

[ghstack-poisoned]

Commit:
Update on "[quant] Statically quantized LSTM"


Differential Revision: [D24270011](https://our.internmc.facebook.com/intern/diff/D24270011)

[ghstack-poisoned]
